### PR TITLE
smtp: avoid counting last eol in file

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/6023

Describe changes:
- smtp: avoid counting last eol in file (in case of plaintext and not base64)

https://github.com/OISF/suricata-verify/pull/1265
```
SV_BRANCH=pr/1265
```

### CODE REVIEW

File: src/util-decode-mime.c [hunk 1]
```diff
@@ -2660,14 +2666,17 @@ static int TestDataChunkCallback(const uint8_t *chunk, uint32_t len,
...
+        }
         PrintChars(SC_LOG_DEBUG, "CHUNK", chunk, len);
 
         /* Parse each line one by one */
         remainPtr = (uint8_t *) chunk;
         do {
             tok = GetLine(remainPtr, len - (remainPtr - (uint8_t *) chunk),
                     &remainPtr, &tokLen);
-            if (tok != NULL && tok != remainPtr) {
+            if (tok != NULL && tok != remainPtr && remainPtr[-1] == LF) {
```
Comment (jasonish) [inline src/util-decode-mime.c:2679]: Can't say I've ever used a negative array pointer, but it kinda makes sense here.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 14788

File: src/util-decode-mime.c (hunk 1 above)
Comment (catenacyber) [reply src/util-decode-mime.c:2679]: It will make more sense to remove all this code with https://github.com/OISF/suricata/pull/9038 ;-)

Comment (inashivb) [reply src/util-decode-mime.c:2679]: I didn't know they existed in C 🥲

Comment (catenacyber) [reply src/util-decode-mime.c:2679]: As Mark Twain said : 

> They did not know it was impossible so they did it

Comment (victorjulien) [reply src/util-decode-mime.c:2679]: Reading this code I don't (clearly) see how we can be sure `remainPtr[-1]` accesses valid memory here.

File: src/util-decode-mime.c [hunk 2]
```diff
@@ -1609,6 +1600,16 @@ static int ProcessBodyLine(const uint8_t *buf, uint32_t len,
             remaining -= tobuf;
             offset += tobuf;
         }
+        if (ret == MIME_DEC_OK) {
+            ret = ProcessDecodedDataChunk(state->data_chunk, state->data_chunk_len, state);
```
Comment (victorjulien) [inline src/util-decode-mime.c:1604]: what does failure mean here for the memcpy below? Should we exit the function here?

Comment (catenacyber) [reply src/util-decode-mime.c:1604]: I think that previous failure should not prevent to save the end of line into the buffer for future addition

Failure of `ProcessDecodedDataChunk` is unimportant (just debug log)

Failure due to `DATA_CHUNK_SIZE - state->data_chunk_len` should be unreachable

Comment (victorjulien) [reply src/util-decode-mime.c:1604]: we do pass the error to the caller, right?

File: src/util-decode-mime.c (hunk 1 above)
Comment (catenacyber) [reply src/util-decode-mime.c:2679]: > Reading this code I don't (clearly) see how we can be sure remainPtr[-1] accesses valid memory here.

This is just testing code (this is not used by Suricata on live traffic, and not even compiled if you do not take the unit tests), and the tests pass :-p

(By the way, these tests are responsible for having a function pointer, hence the big vulnerability that got fixed around)

I guess I should get rid of the call to `GetLine` and count `\n` more basically...

File: src/util-decode-mime.c (hunk 2 above)
Comment (catenacyber) [reply src/util-decode-mime.c:1604]: And the callers just do debug logging about it, right ?

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/9101

**Pull request closed without merging**